### PR TITLE
2911 sequence number virtualization poc

### PR DESCRIPTION
**Pull request by kian-thompson**
# [AB#2911](https://dev.azure.com/fluidframework/internal/_workitems/edit/2911)

**TODO**
- `VirtualDeltaManager` is currently an extension of `DeltaManager`. It might be tricky to make it an adapter instead
- `VirtualContainerStorageAdapter` should be removed in favor of a new wrapper class that contains `ContainerStorageAdapter` as well as other necessary methods for runtime
- Make properties within DeltaManager uniform (naming and ensure usage between real and virtual is not mixed)
- double check no extra virtualization is needed on ".protocol" tree

### CODE REVIEW

File: packages/loader/container-loader/src/deltaManager.ts [hunk 1]
```diff
@@ -263,7 +276,17 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
             assert(batch[batch.length - 1].metadata?.batch === false, 0x3cb /* no end batch markup */);
         }
 
-        this.connectionManager.sendMessages(batch);
+        // TODO: revisit
+        if (batch.length >= 2) {
+            this.connectionManager.sendMessages([{
+                type: GroupedBatchOpType,
+                clientSequenceNumber: batch[0].clientSequenceNumber, // TODO: doesn't matter what we put here because we forget this message
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:283]: your comment is not entirely correct, as protocol cares about clientSequenceNumbers - there should be no gaps. For that reason, we either can't assign clientSequenceNumber to each message in the batch (as that will create gaps when observing ops on the wire), OR we have to virtualize them

File: packages/loader/container-loader/src/deltaManager.ts [hunk 2]
```diff
@@ -263,7 +276,17 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
...
+                referenceSequenceNumber: batch[0].referenceSequenceNumber, // TODO: doesn't matter what we put here because we forget this message
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:284]: Somewhat similar to above: While loader assigns it (and this number is not visible to local client directly), it needs to be "right" (from runtime POV) in a sense of properly referring to a state of container at time this op was made. I.e. it needs to be virtualized

File: packages/loader/container-loader/src/deltaManager.ts [hunk 3]
```diff
@@ -783,12 +812,36 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
...
+                    for (const subMessage of message.contents as IDocumentMessage[]) {
+                        this.updateLatestKnownOpSeqNumber(++this.virtualSequenceNumber);
+                        this._inbound.push({
+                            ...message, // This will override the property difference between ISequencedDocumentMessage and IDocumentMessage
+                            ...subMessage,
+                            sequenceNumber: this.virtualSequenceNumber,
+                        });
+                    }
+                } else {
+                    this.updateLatestKnownOpSeqNumber(++this.virtualSequenceNumber);
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:842]: It will be very hard to follow code and make it right if you start mixing virtualized and non-virtualized values here.
At a minimum, I'd suggest starting using branded types, such that it's impossible to accidently mix and match them and provide conversion functions between types (you will need conversion back and forth - you need herr to map referenceSequenceNumber, clientSequenceNumber as well as sequenceNumber to virtualized numbers).

But I'd suggest using Delta Manager adapter approach. Leveraging branded types on top of it would be a plus, but maybe it's secondary step

Comment (vladsud) [reply packages/loader/container-loader/src/deltaManager.ts:842]: Even if you make it work by injecting code here and there across DM, it will be very hard to follow this code / understand - the risk of future developers making mistakes would increase substantially. An adapter allows isolation / testability, and results in most developers to not need to touch that code ever.

File: packages/loader/container-loader/src/deltaManager.ts [hunk 4]
```diff
@@ -168,26 +173,34 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
...
     public get lastKnownSeqNumber() {
+        // This variable now refers to virtual sequence number
         return this.lastObservedSeqNumber;
     }
 
     public get referenceTerm(): number {
         return this.baseTerm;
     }
 
+    // TODO: possibly deprecate
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:203]: I think you want to have some assessment (as part of these comments) on how possible it is.
I think MSN is used all over the place, so that one is not really feasible. 
In other places I'd ask the key question - what do we gain by trying to deprecate it? For example, I think there is no gain of deprecating initialSequenceNumber - that's sequence number of snapshot, runtime has to know virtualized sequence number one way or another. Or maybe other way to say it - for any op that follows, we have to have capability to take sequence number, reference sequence number and virtualize it. And if we have that capability - it's trivial to provide virtualized initialSequenceNumber

File: packages/loader/container-loader/src/deltaManager.ts (hunk 1 above)
Comment (kian-thompson) [reply packages/loader/container-loader/src/deltaManager.ts:283]: I thought the client sequence number gaps are checked by client? Does the service look at it and do some validation, or is it just treated as some metadata? Or does it matter from the perspective of the connection manager (since it looks to be keeping track)?

File: packages/loader/container-loader/src/deltaManager.ts (hunk 2 above)
Comment (kian-thompson) [reply packages/loader/container-loader/src/deltaManager.ts:284]: My thinking was this new op is essentially ignored once DeltaManager gets the ack for it. It recognizes it and decomposes to the original ops with virtual sequence numbers (which contain the virtualized reference sequence number from when it was submitted).

File: packages/loader/container-loader/src/deltaManager.ts (hunk 1 above)
Comment (vladsud) [reply packages/loader/container-loader/src/deltaManager.ts:283]: Currently there are two purposes for clientSequenceNumber concept:
- service validates that all messages for given connection have consecutive numbers
- client uses them to identify messages (i.e. as identity mechanism)

File: packages/loader/container-loader/src/virtualDeltaManager.ts [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+        }
+        this._trackedListeners.clear();
+    }
+
+    public peek(): T | undefined {
+        const value = this._queue.peek();
+        return value !== undefined ? this._virtualizeValue(value) : undefined;
+    }
+
+    public toArray(): T[] {
```
Comment (vladsud) [inline packages/loader/container-loader/src/virtualDeltaManager.ts:60]: Is this method used? can we remove it / deprecate it?

File: packages/loader/container-loader/src/virtualDeltaManager.ts [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+
+    public get idle(): boolean {
+        return this._queue.idle;
+    }
+
+    public get disposed(): boolean {
+        return this._queue.disposed;
+    }
+
+    public async pause(): Promise<void> {
```
Comment (vladsud) [inline packages/loader/container-loader/src/virtualDeltaManager.ts:85]: Can you use DeltaQueueProxy as a base class here and avoid overwriting methods that are pass-through?

File: packages/loader/container-loader/src/virtualDeltaManager.ts [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+        super();
+
+        this.on("newListener", (event, _listener) => {
+            assert((this.listeners(event).length !== 0) === this._trackedListeners.has(event), "mismatch");
+            if (!this._trackedListeners.has(event)) {
+                let listener: (...args: any[]) => void;
+
+                // eslint-disable-next-line unicorn/prefer-ternary
+                if (event === "push" || event === "op") {
+                    listener = (task: T) => { this.emit(event, this._virtualizeValue(task, true /* storeValue */)); }
```
Comment (vladsud) [inline packages/loader/container-loader/src/virtualDeltaManager.ts:32]: General note: it's easy to reason about what needs to happen when direction of communication is clear. I.e. runtime pushes ops into the queue.
It's hard to impossible when
- direction is unknown (for example, "push" - is this result of runtime pushing (indirectly) into outbound queue, or it's connection manager pushing received ops into inbound queue
- methods like toArray, where we do not know anything about consumer (and direction of the flow).

So I have a feeling that doing transformations here is wrong. It's very hard to ensure that we covered all the ways someone can access content (i.e. why do we convert here on "op" but not on peek()?)

Plus "push" for inbound & outbound queues needs different type of transformation, right?

Instead, I think we should push responsibility to do conversions to next layer (DM wrapper) when content is pushed into the queue, or when content is consumed from the queue.

File: packages/loader/container-loader/src/virtualDeltaManager.ts [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+    extends DeltaManager<TConnectionManager> {
+
+    private _virtualSequenceNumber = 0;
+    private _clientSequenceNumber = 0;
+
+    /**
+     * Key: real client sequence number
+     * Value: virtual client sequence number
+     */
+    private readonly _clientSequenceNumberMap = new Map<number, number>();
```
Comment (vladsud) [inline packages/loader/container-loader/src/virtualDeltaManager.ts:108]: Do we need this map? I.e. we talked about stop projecting clientSequenceNubmers to runtime, and then probably we do not need it?

File: packages/loader/container-loader/src/virtualDeltaManager.ts [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+            clientSequenceNumber: this.connectionManager.clientId === message.clientId ? this.virtualizeClientSequenceNumber(message.clientSequenceNumber) : message.clientSequenceNumber,
+        };
+    }
+
+    // TODO: need to review how these numbers are being used in other layers
+    // there is currently some mix and match happening
+
+    // TODO: virtualize this number
+    public get initialSequenceNumber(): number {
+        return this.initSequenceNumber;
```
Comment (vladsud) [inline packages/loader/container-loader/src/virtualDeltaManager.ts:260]: Isn't this
return this.virtualizeSequenceNumber(this.initSequenceNumber);

?

File: packages/loader/container-loader/src/virtualDeltaManager.ts (hunk 7 above)
Comment (vladsud) [reply packages/loader/container-loader/src/virtualDeltaManager.ts:32]: Oh, I realized that one point is not valid. How about changing _virtualizeValue work to "transform"? As it can go both directions.

File: packages/loader/container-loader/src/virtualDeltaManager.ts [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+    public async attachOpHandler(
+        minSequenceNumber: number,
+        sequenceNumber: number,
+        term: number,
+        handler: IDeltaHandlerStrategy,
+        prefetchType: "cached" | "all" | "none" = "none",
+    ) {
+        this._virtualSequenceNumber = sequenceNumber;
+
+        this._sequenceNumberMap.set(sequenceNumber, sequenceNumber);
```
Comment (vladsud) [inline packages/loader/container-loader/src/virtualDeltaManager.ts:162]: initialization seems non-complete (for prototype).
I think you want to init all the numbers in teh range of MSN...sequenceNumber, as you will observe ops with referenceSequenceNumber from other clients in between of these low & hi bounds.

File: packages/loader/container-loader/src/virtualDeltaManager.ts [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+        this._clientSequenceNumber = 0;
+        super.connectHandler(connection);
+    }
+
+    protected sendMessageToConnectionManager(messages: IDocumentMessage[]): void {
+        if (messages.length >= 2) {
+            this.connectionManager.sendMessages([{
+                type: GroupedBatchOpType,
+                clientSequenceNumber: ++this._clientSequenceNumber,
+                referenceSequenceNumber: messages[0].referenceSequenceNumber, // use oldest referenceSequenceNumber
```
Comment (vladsud) [inline packages/loader/container-loader/src/virtualDeltaManager.ts:180]: We should assert here that all refereceSequenceNumbers are the same.
We have same dependency for compression, so you should be able to ride on his work (and validation), but it's better to assert assumptions in the code.

File: packages/loader/container-loader/src/virtualDeltaManager.ts [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+                    clientSequenceNumber: subMessage.clientSequenceNumber, // This number is already virtualized
+                    referenceSequenceNumber: this.virtualizeSequenceNumber(subMessage.referenceSequenceNumber),
+                    minimumSequenceNumber: this.virtualizeSequenceNumber(message.minimumSequenceNumber),
+                }, startTime);
+            }
+        } else {
+            super.processMessage({
+                ...message,
+                sequenceNumber: this.virtualizeSequenceNumber(message.sequenceNumber),
+                clientSequenceNumber: this.connectionManager.clientId === message.clientId ? this.virtualizeClientSequenceNumber(message.clientSequenceNumber) : message.clientSequenceNumber,
```
Comment (vladsud) [inline packages/loader/container-loader/src/virtualDeltaManager.ts:210]: This does not look right. Tracking code, this.connectionManager.clientId is current connection ID. But we could get our own ops (acks) from previous connection while already on new connection. This would be challenging to do right at this layer, as only Container (and ConnectionStateHandler) know about that extra layer of "virtualization" of clientSequence numbers.

Which actually raises an interesting question: there is no abstraction here, i.e. these numbers leak to other layers. We attempt to patch only them for our own ops. But there could be assumptions somewhere in the code that uses clientSequenceNumber for any op to do some kind of sorting or something else. So we would either need to keep a map for all clients (which is almost impossible, as we lack info when we join for the first time), OR we will need to store that virtualized nubmers on ops themselves, OR we need to get rid of clientSequenceNumber on the whole interface (ISequencedDocumentMessage)

Comment (vladsud) [reply packages/loader/container-loader/src/virtualDeltaManager.ts:210]: For prototype, you can start with simple - add it to payload (content), such that all clients have protocol and runtime clientSequenceNumber (for first message in the group). Ugly, but will do the work

File: packages/loader/container-loader/src/virtualDeltaManager.ts [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+    public get initialSequenceNumber(): number {
+        return this.initSequenceNumber;
+    }
+
+    public get lastSequenceNumber(): number {
+        return this.virtualizeSequenceNumber(this.lastProcessedSequenceNumber);
+    }
+
+    public get lastMessage() {
+        return this.lastProcessedMessage !== undefined ? this.virtualizeSequencedMessage(this.lastProcessedMessage) : undefined;
```
Comment (vladsud) [inline packages/loader/container-loader/src/virtualDeltaManager.ts:268]: Isn't it double-virtualizaiton? I.e. this.lastProcessedMessage content is result of inbound queue processing.

File: packages/loader/container-loader/src/virtualDeltaManager.ts [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+        } else {
+            this.connectionManager.sendMessages(messages.map((it: IDocumentMessage) => {
+                this._clientSequenceNumberMap.set(++this._clientSequenceNumber, it.clientSequenceNumber);
+                return {...it, clientSequenceNumber: this._clientSequenceNumber};
+            }));
+        }
+    }
+
+    protected processMessage(message: ISequencedDocumentMessage, startTime: number): void {
+        if (message.type === GroupedBatchOpType) {
```
Comment (vladsud) [inline packages/loader/container-loader/src/virtualDeltaManager.ts:192]: Maybe I'm wrong, but I feel like you are doing double-virtualization - first one when op is pushed into inbound queue, and second one when it is processed here.

File: packages/loader/container-loader/src/virtualDeltaManager.ts [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+        }
+        const virtualizedClientSequenceNumber = this._clientSequenceNumberMap.get(clientSequenceNumber);
+        assert(virtualizedClientSequenceNumber !== undefined, "clientSequenceNumber not found");
+
+        return virtualizedClientSequenceNumber;
+    }
+
+    private virtualizeSequencedMessage(message: ISequencedDocumentMessage, storeValue: boolean = false): ISequencedDocumentMessage {
+        if (storeValue) {
+            if (message.type === GroupedBatchOpType) {
```
Comment (vladsud) [inline packages/loader/container-loader/src/virtualDeltaManager.ts:239]: I understand why you are going with inheritance (some things are easier that way). But I think it will not work (the way it is right now) :(
Once "group op" is in the queue, its visible to runtime. There is a leak across virtualized / non-virtualized boundary.

I think you will need to remove access to inbound queue for "real" DM and abstract it away, such that when op is coming in, some method is called and that's when you expand it and push it into "virtualized" queue individual ops.
Probably something similar needs to be done for outbound? Not sure.

Basically, there is a lot of leakage in our current interfaces. Ideally queues should not be exposed.
And I think we can do it, I just do not know (yet) how to order work (and if we should focus on not changing interfaces yet by making more abstractions at DM level instead).
I.e. with these changes, I believe we can get rid of ScheduleManager class and that's probably the only place where we peek into the queues. And with that we can probably hide queues from IDM interface (exposing only functionality that is needed by runtime, like op pausing). This will have impact on back compat, but I'd not worry about it yet - we can figure it out, I'm sure!

Let's chat about it.

File: packages/loader/container-loader/src/virtualDeltaManager.ts (hunk 5 above)
Comment (kian-thompson) [reply packages/loader/container-loader/src/virtualDeltaManager.ts:60]: This is being used by the ScheduleManager ctor

File: packages/loader/container-loader/src/virtualDeltaManager.ts (hunk 6 above)
Comment (kian-thompson) [reply packages/loader/container-loader/src/virtualDeltaManager.ts:85]: I'm not sure I can since DeltaQueueProxy forwards events whereas I want to intercept them and virtualize before re-emitting

File: packages/loader/container-loader/src/virtualDeltaManager.ts [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+        if (messages.length >= 2) {
+            this.connectionManager.sendMessages([{
+                type: GroupedBatchOpType,
+                clientSequenceNumber: ++this._clientSequenceNumber,
+                referenceSequenceNumber: messages[0].referenceSequenceNumber, // use oldest referenceSequenceNumber
+                contents: messages,
+            }]);
+        } else {
+            this.connectionManager.sendMessages(messages.map((it: IDocumentMessage) => {
+                this._clientSequenceNumberMap.set(++this._clientSequenceNumber, it.clientSequenceNumber);
```
[the remark this answers is not in the document]
Comment (kian-thompson) [reply packages/loader/container-loader/src/virtualDeltaManager.ts:185]: The grouped messages already contain the virtualized clientSequenceNumber, since the parent op would have the "real" clientSequenceNumber (but we don't send it to the other layers)

File: packages/loader/container-loader/src/virtualDeltaManager.ts (hunk 13 above)
Comment (kian-thompson) [reply packages/loader/container-loader/src/virtualDeltaManager.ts:268]: Not technically. The _inbound queue stores the un-virtualized values. It's only when other places try to ask for those values where they get virtualized. All the values are un-virtualized in DeltaManager

File: packages/loader/container-loader/src/virtualDeltaManager.ts (hunk 14 above)
Comment (kian-thompson) [reply packages/loader/container-loader/src/virtualDeltaManager.ts:248]: Same note as above
> The _inbound queue stores the un-virtualized values. It's only when other places try to ask for those values where they get virtualized. All the values are un-virtualized in DeltaManager

File: packages/loader/container-loader/src/deltaManager.ts [hunk 17]
```diff
@@ -388,7 +392,7 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
             "connect",
             connection,
             checkpointSequenceNumber !== undefined ?
-                this.lastObservedSeqNumber - this.lastSequenceNumber : undefined);
+                this.lastObservedSeqNumber - this.lastProcessedSequenceNumber : undefined);
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:395]: Maybe it's worth (as separate PR) rename these, for easier reading. Ideally, we follow this schema - lastSequenceNumber  (public getter) - _lastSequenceNumber  (private)
That said, I think it's not very safe to rely on a fact here (I think that's the case) that they have vastly different meaning (virtualized vs. not) - developers who fix bugs in this code will accidently use these interchangeably, unless we have really good test coverage to catch it (essentially some test that validates public getters are not called form within this class by having derived test class that catches such things)

File: packages/loader/container-loader/src/virtualDeltaManager.ts [hunk 18]
```diff
@@ -0,0 +1,335 @@
...
+        this.on("newListener", (event, _listener) => {
+            assert((this.listeners(event).length !== 0) === this._trackedListeners.has(event), "mismatch");
+            if (!this._trackedListeners.has(event)) {
+                let listener: (...args: any[]) => void;
+
+                // eslint-disable-next-line unicorn/prefer-ternary
+                if (event === "push" || event === "op") {
+                    listener = (task: T) => {
+                        // Let VirtualDeltaManager know it should update sequence number map (for inbound)
+                        _storeVirtualizedSequenceNumber?.(task);
```
Comment (vladsud) [inline packages/loader/container-loader/src/virtualDeltaManager.ts:46]: nit: you can always provide empty callback - I think code becomes a bit cleaner with it.

I'd name things in context of this class and expectations. I feel like it should take the following form:
observe: (value: Readonly<T>) => void

File: packages/loader/container-loader/src/virtualDeltaManager.ts [hunk 19]
```diff
@@ -0,0 +1,335 @@
...
+    ) {
+        super();
+
+        this.on("newListener", (event, _listener) => {
+            assert((this.listeners(event).length !== 0) === this._trackedListeners.has(event), "mismatch");
+            if (!this._trackedListeners.has(event)) {
+                let listener: (...args: any[]) => void;
+
+                // eslint-disable-next-line unicorn/prefer-ternary
+                if (event === "push" || event === "op") {
```
Comment (vladsud) [inline packages/loader/container-loader/src/virtualDeltaManager.ts:43]: Can we deprecate these events (and access to queues)?
In a context of dealing with back compat later (i.e. assume that runtime and loader are coming from your PR only)?
I'd keep ability to pause queues (as it's hard to get rid of it) as a stand-alone API on IDM interface, but can we otherwise get rid of the queue exposure on public API? Is the only user here ScheduleManager? If so, I can probably help by prototyping (on top of your PR) if we can remove it (or its need to poke at these events)

File: packages/loader/container-loader/src/virtualDeltaManager.ts [hunk 20]
```diff
@@ -0,0 +1,335 @@
...
+            return undefined;
+        }
+        // Decompose potential "groupedBatch" and send the last message of that grouping
+        const valueArr = this._virtualizeValue(value);
+        return valueArr[valueArr.length - 1];
+    }
+
+    public toArray(): T[] {
+        const virtualizedArr: T[] = [];
+        this._queue.toArray().forEach((value) => {
```
Comment (vladsud) [inline packages/loader/container-loader/src/virtualDeltaManager.ts:85]: Is this same as this._queue.toArray().map((value) => this._virtualizeValue(value)) ?

File: packages/loader/container-loader/src/virtualDeltaManager.ts (hunk 6 above)
Comment (vladsud) [reply packages/loader/container-loader/src/virtualDeltaManager.ts:112]: Sorry, I meant DeltaManagerProxy.

File: packages/loader/container-loader/src/virtualDeltaManager.ts [hunk 21]
```diff
@@ -0,0 +1,335 @@
...
+ * Key point is that anyone who wants to observe the events or see the content of the DeltaQueue
+ * will see the version that is virtualized (not what is sent over the wire).
+ *
+ * For example:
+ * - Say we have ops 1, 2, 3 and they get grouped together in a new "groupedBatch" op
+ * - Over the wire, we send just one op and the DeltaQueue will store just this single op
+ * - When other places subscribe to "push"/"op" event or call "toArray"/"peek", they will see a version that
+ * is inline with when the ops 1, 2, 3 were separate ops and will not see the "groupedBatch" op
+ */
+class VirtualDeltaQueueProxy<T> extends TypedEventEmitter<IDeltaQueueEvents<T>> implements IDeltaQueue<T> {
```
Comment (vladsud) [inline packages/loader/container-loader/src/virtualDeltaManager.ts:27]: Overall observation RE exposure of queues & leakage:
1. Looking at usage of inboud queue in DM, I think we can remove it from there by adding couple abstract members, and not implementing IDQ interface (i.e. pushing implementation of public interface to be only on virtualized class). That way, inbound is projected only down (to runtime), and it's semantics are very clear. 
2. outboud is tough. It's implemented by ConnectionManager. DM never touches it in any way. But I think we can get rid of this concept altogether in ConnectionManager. Basically, queue should be only used when we have a connection. And in such case, it's not really a queue, as messages are pushed right away to socket. It's async today (queue processes items async, so it does work a bit like a queue, i.e. it could accumulate items), but it does not have to - there are no issues with sending ops right away (same can't be said for inboud queue!). I wonder, if we could prototype getting rid of outbound queue concept in ConnectionManager as a separate PR?
   - I see the only place where we use it in runtime - see connectionTelemetry.ts. And in there, we assume we see non-processed ops right before we send them. This telemetry measures how long it takes from op generation (on runtime side) till it actually gets sent. But with batching accumulation on runtime now, we should be able to rely instead on ContainerRuntime.flush() providing these signals

Comment (vladsud) [reply packages/loader/container-loader/src/virtualDeltaManager.ts:27]: One addition - it's a a lot of processing (packing/unpacking) for "push" flow on oubound queue. It's very inefficient (from memory consumption / CPU consumption POV), so avoiding it (by ensuring runtime does not need to rely on it) would be great.
It should be possible to change that in separate PR - I'd not try to jam everything into here (and it's good change to make regardless). Please let me know if you need help - I'm more familiar with that code and can help.

File: packages/runtime/container-runtime/src/pendingStateManager.ts [hunk 22]
```diff
@@ -218,21 +214,6 @@ export class PendingStateManager implements IDisposable {
         assert(pendingState.type === "message", 0x169 /* "No pending message found for this remote message" */);
         this.pendingStates.shift();
 
-        // Processing part - Verify that there has been no data corruption.
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/pendingStateManager.ts:221]: consider splitting these changes into separate PR and get it in sooner (to keep your main PR smaller).

Comment (kian-thompson) [reply packages/runtime/container-runtime/src/pendingStateManager.ts:221]: https://github.com/microsoft/FluidFramework/pull/13917

File: examples/data-objects/codemirror/src/index.ts [hunk 23]
```diff
@@ -72,7 +72,9 @@ class CodeMirrorFactory extends RuntimeFactoryHelper {
             buildRuntimeRequestHandler(
                 mountableViewRequestHandler(MountableView, [viewRequestHandler]),
             ),
-            undefined, // runtimeOptions
+            {
+                summaryOptions: { summaryConfigOverrides: { state: "disabled" } },
```
Comment (vladsud) [inline examples/data-objects/codemirror/src/index.ts:76]: Should this be removed, given you mentioned summaries work now?

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 24]
```diff
@@ -2865,8 +2861,11 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
             assert(!!maybeSnapshot, 0x138 /* "Failed to get snapshot from storage" */);
             stats.getSnapshotDuration = trace.trace().duration;
 
+            const attributes = await this.storage.getDocumentAttributesFromTree?.(maybeSnapshot);
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2864]: I'm a bit confused here RE why we need new method on storage interface, and what is the meaning of sequence number s in there. Drivers, for most part, should not have knowledge about attributes and treat is as any other blob.
I think today attributes concept is a loader concept (though storage knows about it for sure), and ideally it should stay that way, unless we want to expose non-virtualized sequence numbers to runtime.

Can you please share a bit more on what you are trying to accomplish here?

File: examples/data-objects/codemirror/src/index.ts (hunk 23 above)
Comment (kian-thompson) [reply examples/data-objects/codemirror/src/index.ts:76]: Didn't mean to commit this as it is only relevant to testing
I haven't done example testing with summaries yet, but the main areas should be covered

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 24 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2864]: So the issue here is that code has access to protocol tree that has non-virtualized data. I.e. leakage.
I think we have to figure out how to prevent that leakage and/or patch data in a way where no runtime changes are required.
ODSP driver used to do it (there is still code artifacts from those days, I believe). I think loader layer (virtual storage layer) has to do it, by creating "virtual blob" and patching tree with updated .attributes blob.
That has some negative implications (we introduce fake blob IDs, and theoretically runtime can reuse them). That said, it's easier to implement and its overall pretty solid, so we can start here and consider switching to other design (with more invasive changes) later.

Maybe better way - remove protocol tree from tree when passing it to runtime. And provide explicit APIs to get this data. What if we modify ISnapshotTree itself to include this data directly? I.e. introduce ISnapshotTreeRoot that extends ISnapshotTree and has decoded data from protocol, but omits protocol tree itself? That way loader can patch it while providing to runtime.

File: packages/common/driver-definitions/src/storage.ts [hunk 25]
```diff
@@ -177,6 +178,11 @@ export interface IDocumentStorageService extends Partial<IDisposable> {
      * server has deleted it this call may result in a broken promise.
      */
     downloadSummary(handle: ISummaryHandle): Promise<ISummaryTree>;
+
+    /**
+     * TODO: left as optional for now to keep implementing classes cleaner
+     */
+    getDocumentAttributesFromTree?(tree: ISnapshotTree): Promise<IDocumentAttributes>;
```
Comment (vladsud) [inline packages/common/driver-definitions/src/storage.ts:185]: I do not think this belongs here. From driver implementation perspective, that can be implemented as stand-alone function, so it does not belong on this interface.
If we really want to, we can always have different interface that we expose to runtime that does implement this API. But its presence is an indication of leakage - we leak non-virtualized data in the form of .attribute blob in a protocol tree. Exposing this method does not fix it (someone not knowing about this API would still get to wrong data).
See my other comments in runtime around usage on how we might approach it.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 24 above)
Comment (kian-thompson) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2864]: Currently, the runtime-utils is doing this document attributes extraction (see runtime-utils/src/utils.ts).

I agree, the surface can probably be simplified in terms of the runtime not reaching into the tree for the values and instead being provided them explicitly. I think doing it in a more explicit way is safer than modifying the tree because it does not require sync on both sides (loader and runtime) in terms of the structure of the object.

**Comment by kian-thompson** [comment]
Closing in favor of the grouped batching solution https://github.com/microsoft/FluidFramework/pull/14512
